### PR TITLE
refactor(connector): do not expose internal implementation for parser benchmarks

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As described in #16968,

> Since https://github.com/risingwavelabs/risingwave/pull/13707, `JsonParser`(`src/connector/src/parser/json_parser.rs`) has been superseded by `PlainParser` to parse `FORMAT PLAIN ENCODE JSON`.

This silently breaks the benchmark like `parser` and `nexmark_integration` as they are still operating on `JsonParser` and cannot reflect the real code path anymore.

To resolve this problem, this PR refactors them to use outer-most interface `ByteStreamSourceParserImpl` instead, allowing the internal implementation to be arbitrarily changed in the future.

This will be a step towards #16968. The only usages of `JsonParser` are now:

- The benchmark `json_vs_plain_parser`
- Unit tests in `json_parser.rs`

In following PRs, we can refactor the unit tests to depend on `JsonAccessor` or `PlainParser`. Then we can move `JsonParser` into the benchmark `json_vs_plain_parser` and make it a private test-only component.

Also refactor the structure of the benchmarks:

- Split `json_parser` to `debezium_json_parser` and `json_vs_plain_parser`. Extract common codes.
- Rename `parser` to `json_parser_case_insensitive`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by xxchan — APPROVED** [review]

File: src/connector/benches/nexmark_integration.rs [hunk 1]
```diff
@@ -86,9 +91,12 @@ fn make_parser() -> impl ByteStreamSourceParser {
     .map(|(i, (n, t))| SourceColumnDesc::simple(n, t, ColumnId::new(i as _)))
     .collect_vec();
 
-    let props = SpecificParserConfig::DEFAULT_PLAIN_JSON;
+    let config = ParserConfig {
+        common: CommonParserConfig { rw_columns },
```
Comment (stdrc) [inline src/connector/benches/nexmark_integration.rs:95]: Interesting, why `rw_columns` here?

File: src/connector/src/parser/mod.rs [hunk 2]
```diff
@@ -937,6 +937,11 @@ impl ByteStreamSourceParserImpl {
             _ => unreachable!(),
         }
     }
+
+    /// Create a parser for testing purposes.
+    pub fn create_for_test(parser_config: ParserConfig) -> ConnectorResult<Self> {
```
Comment (stdrc) [inline src/connector/src/parser/mod.rs:942]: Why not `new_for_test`?

File: src/connector/src/source/base.rs [hunk 3]
```diff
@@ -543,6 +543,19 @@ pub struct SourceMessage {
     pub meta: SourceMeta,
 }
 
+impl SourceMessage {
+    /// Create a dummy `SourceMessage` with all fields unset for testing purposes.
+    pub fn dummy() -> Self {
```
Comment (stdrc) [inline src/connector/src/source/base.rs:548]: Is the `SourceContext::dummy` still in use?

**Review by stdrc — APPROVED** [review]

File: src/connector/benches/nexmark_integration.rs (hunk 1 above)
Comment (BugenZhao) [reply src/connector/benches/nexmark_integration.rs:95]: Not sure. Leave it untouched.

File: src/connector/src/parser/mod.rs (hunk 2 above)
Comment (BugenZhao) [reply src/connector/src/parser/mod.rs:942]: To be consistent with the method above (`create`).

File: src/connector/src/source/base.rs (hunk 3 above)
Comment (BugenZhao) [reply src/connector/src/source/base.rs:548]: Yes.

**Pull request merged**